File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def __init__(self):
        self.vec = []
    def inorder(self, root):
        if root is None:
            return
        self.inorder(root.left)
        self.vec.append(root.val)
        self.inorder(root.right)
        return

    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None:
            return True
        self.inorder(root)
        for i in range(1,len(self.vec)):
            if self.vec[i] <= self.vec[i-1]:
                return False
        return True

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* prev = nullptr; // record previous node

    bool isValidBST(TreeNode* root) {
        if (!root)
            return true;        
        bool leftbool = isValidBST(root->left);
        if (prev && prev->val >= root->val) return false;
        else prev = root;
        bool rightbool = isValidBST(root->right);
        return leftbool && rightbool;
    }
};